pkg/datapath/tables/node_addressing.go

```diff
@@ -0,0 +1,141 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package tables
+
+import (
+	"context"
+	"net"
+
+	"github.com/cilium/cilium/pkg/cidr"
+	"github.com/cilium/cilium/pkg/datapath/types"
+	"github.com/cilium/cilium/pkg/hive/cell"
+	"github.com/cilium/cilium/pkg/node"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/cilium/cilium/pkg/statedb"
+)
+
+// NodeAddressingCell provides the [NodeAddressing] interface that provides
+// access to local node addressing information. This will be eventually
+// superceded by Table[NodeAddress].
+var NodeAddressingCell = cell.Module(
+	"node-addressing",
+	"Accessors for looking up local node IP addresses",
+
+	cell.Provide(NewNodeAddressing),
+)
+
+func NewNodeAddressing(localNode *node.LocalNodeStore, db *statedb.DB, nodeAddresses statedb.Table[NodeAddress], devices statedb.Table[*Device]) types.NodeAddressing {
+	return &nodeAddressing{
+		localNode:     localNode,
+		db:            db,
+		nodeAddresses: nodeAddresses,
+		devices:       devices,
+	}
+}
+
+type nodeAddressing struct {
+	localNode     *node.LocalNodeStore
+	db            *statedb.DB
+	nodeAddresses statedb.Table[NodeAddress]
+	devices       statedb.Table[*Device]
+}
+
+func (n *nodeAddressing) IPv6() types.NodeAddressingFamily {
+	return addressFamily{n, ipv6}
+}
+
+func (n *nodeAddressing) IPv4() types.NodeAddressingFamily {
+	return addressFamily{n, ipv4}
+}
+
+func (a addressFamily) Router() net.IP {
+	if n, err := a.localNode.Get(context.Background()); err == nil {
+		return n.GetCiliumInternalIP(a.flags&ipv6 != 0)
+	}
+	return nil
+}
+
+func (a addressFamily) PrimaryExternal() net.IP {
+	if n, err := a.localNode.Get(context.Background()); err == nil {
+		return n.GetNodeIP(a.flags&ipv6 != 0)
+	}
+	return nil
+}
+
+func (a addressFamily) AllocationCIDR() *cidr.CIDR {
+	if n, err := a.localNode.Get(context.Background()); err == nil {
+		return n.IPv4AllocCIDR
+	}
+	return nil
+}
+
+func (a addressFamily) LocalAddresses() (addrs []net.IP, err error) {
+	return a.getNodeAddresses(a.db.ReadTxn(), a.flags), nil
+}
+
+func (a addressFamily) DirectRouting() (int, net.IP, bool) {
```

Comment:
This is likely half-implemented with not all uses of `option.Config.DirectRoutingDevice` converted.